README.md
Overview
========
LEDscape is a library and service for controlling individually addressable LEDs from a 
Beagle Bone Black or Beagle Bone Green using the onboard [PRUs](http://processors.wiki.ti.com/index.php/Programmable_Realtime_Unit_Subsystem). It currently supports WS281x 
(WS2811, WS2812, WS2812b), WS2801, sk6812 and initial support for DMX. 

It can support up to 48 connected strings of WS281x and 12 strings of sk6812x LED pixels *including RGBW pixels* and can drive them with very little load on the main processor.

Background
------
LEDscape was originally written by Trammell Hudson (http://trmm.net/Category:LEDscape) for
controlling WS2811-based LEDs. Since his original work, his version (https://github.com/osresearch/LEDscape)
has been repurposed to drive a different type of LED panel (e.g. http://www.adafruit.com/products/420).

This version of the library was forked from his original WS2811 work. Various improvements have been made in the 
attempt to make an accessible and powerful LED driver based on the BBB. Many thanks to Trammell for his excellent work
in scaffolding the BBB and PRUs for driving LEDs.


Installation
=====================================
It is necessary to have access to a shell onto the Beaglebone using serial, Ethernet, or USB connections.  
Examples on how to do this can be found at [BeagleBoard.org](http://beagleboard.org/getting-started) or at
[Adafruit's Learning Site](https://learn.adafruit.com/ssh-to-beaglebone-black-over-usb/ssh-on-mac-and-linux).

###Start with a compatible Linux image

To use LEDscape, you must use a version of the Linux kernel that supports the `uio_pruss` module. The Beaglebone.org Wheezy or Jessie Linux images work well. 

####Checking existing Linux version

Check which Debian version you are currently running by entering...

```
cat /etc/debian_version
```

This `README` was tested with version `7.11`, but any 7.x version should work. Version 8.x is also compatible as long as  kernel such as the 4.4.x-ti kernel which is built with the `uio_pruss` module is used as opposed to the `remoteproc` PRUSS interface. 

####Installing a compatible Linux version

If you have an incompatible version currently installed or just want to start with a clean install of to the most recent compatible version, you can follow the instructions here under "Update board with latest software"...

http://beagleboard.org/getting-started


###Installing the LEDscape software (OUTDATED)

Log into a compatible Linux version as root and enter the following commands at a command line...

	git clone git://github.com/nkaminski/LEDscape
	cd LEDscape
	chmod +x install-software.sh
	./install-software.sh
	reboot
	
This will do the following....

1. Clone the LEDscape repository to your local machine in a directory called "LEDscape" under whatever directory you started in.
3. Make the install script executable.
2. Build the LEDscape software from the sources. This takes a couple of minutes and you will see lots of scrolling. 
3. Copy the new flattened device tree files to your `/boot` directory. These files enable the PRU subsystem.  Note that the old files are backed up with the extension `preledscape_bk`.
4. Copy a default config file to `/etc/ledscape-config.json` if that file does not already exist. 
4. Install the `uio_pruss` kernel module to let LEDscape talk to the PRU subsystem. 
5. Reboot the machine.

###Updating an existing install

If you are using an older version of LEDscape that keeps the configuration in a JSON file inside the LEDscape directory, you should copy your modified config to `/etc/ledscape-config.json`.  

You should be able to update an existing install with the above procedure without overwriting your configuration in `/etc/ledscape-config.json`. 

Note that the install process will not preserve any modified pin mappings.
 
###Testing the install

Once the machine has rebooted, log in as root, enter the following commands to switch into the `LEDscape` directory you created above and manually start the LEDscape server...

```
cd LEDscape
./run-ledscape
```

It should print some diagnostic messages and continue running until you press Control-C or logout or reboot. 

By default, the server starts sending a demo pattern in the WS2812B format on the output pins. If you connect the `DI` of some strings to these pins, they should light up.  Pins P8-8, P8-10, P8-12, P8-14, P8-16, and P8-18 are great for testing since they are located near a ground on pin P8-2, they are all right next to each other, and they should always have pixel data in the default configuration. If using the sk6812rgbw-r30 output module, data will output for strips 0-5 on pins P9-31, P9-29, P9-30, P9-28, P9-42 and P9-27 respectively. 


###Setting the server to run automatically as a service

If you want LEDscape to automatically start every time the machine is rebooted, you can install it as a service with the following command (run from a command line inside the LEDscape directory as root)...

	sudo ./install-service.sh
	

Open Pixel Control Server
=========================

##Configuration
	
By default LEDscape is configured for strings of 256 WS2811 pixels, accepting OPC
data on port 7890. You can adjust this by editing `run-ledscape` and 
editing the parameters to `opc-server`


##Data Format

The `opc-server` server accepts data on OPC channel 0. It expects the data for
each LED strip concatenated together. This is done because LEDscape requires
that data for all strips be present at once before flushing data data out to
the LEDs. 

`opc-server` supports both TCP and UDP data packets. The TCP port is specified with `--tcp-port <port>` and the UDP port
with `--udp-port <port>`. Entering `0` for a port number will disable that server.
 
Note that if using the UDP server, `opc-server` will limit the number of pixels to 21835, or 454 pixels per port if
using all 48 ports.

##Output Modes

LEDscape is capable of outputting several types of signal. By default, a ws2811-compatible signal is generated. The
output mode can be specified with the `--mode <mode-id>` parameter. A list of available modes and their descriptions
can be obtained by running `opc-server -h`. 

Frame Rates for WS2812 Leds
-----------
	512 per channel ~= 060 fps
	256 per channel ~= 120 fps
	128 per channel ~= 240 fps
	064 per channel ~= 400 fps

Pin Mappings
------------
Each output mode of LEDscape is compatible with several different pin mappings. These pin-mappings are declared in
`pru/mappings` as json files and each contain information about the mapping. They can be provided to LEDscape with the
`--mapping <mapping-id>` parameter, where `<mapping-id>` is the filename of the json file without it's extension.

The mappings are designed for use with various different cape configurations to simplify the hardware designed.
Additional mappings can be created by adding new `json` files to the `pru/mappings` directory and rebuilding. Please note that due to limitations intrinsic to the BeagleBone Black platform, remapping pins is NOT supported for any x-r30 output modes. 

A human-readable pinout for a mapping can be generated by running

    node pru/pinmap.js --mapping <mapping-id>
    
###Default pin mappings

By default, LEDscape is set up to drive 48 strings of WS2812B LEDs, with each string having up to 600 pixels. You can connect shorter strings with no problems except that the update rate will be slower. If you connect longer strings, only the first 600 pixels will update. 

Here is the default mapping of channels (in green) to pins...

![](default-pin-mapping.PNG)


Here are the default pin assignments for the first 6 channels so you can get your bearings...

![](ws2812b-physical-pins.png)

These are the pins you would connect the to each string's `DI` (Data In).

###HDMI conflict

#### BeagleBone Green
The BeagleBone Green has no HDMI port, so all 48 channels are available on the mapped pins by default. 

In fact, if you try to do the edit below you can make the board unbootable!

#### BeagleBone Black
On the BeagleBone Black, the HDMI port uses pins P8-27 through P8-46 so these channels (basically the bottom half of the right header) will not be usable for pixel data by default. If you need less than 28 channels total, you can just use pins that are not assigned to the HDMI port. Furthermore, if you require more than the 6 outputs from PRU0 when using an x-r30 output mode, this change is also required.

If you want to use the pins assigned to the HDMI port for pixel data, then you will need to disable the HDMI port by entering the command...

```
nano /boot/uEnv.txt
```

Find the lines that say...

```
##BeagleBone Black: HDMI (Audio/Video) disabled:
#dtb=am335x-boneblack-emmc-overlay.dtb
```

...and change them to say...

```
##BeagleBone Black: HDMI (Audio/Video) disabled:
dtb=am335x-boneblack-emmc-overlay.dtb
```

(delete the `#` at the beginning of the second line).

Then save the file by pressing `Control-X` and answering `y`, and then reboot by entering `reboot`.

When the board comes back up, the HDMI should be disabled and the pins available for LEDs. 


Multi-Pin Channels
------------------
Some mappings, such as ws2801, use multiple pins to output each channel. In these cases, fewer than the full 48 channels
are available. In the case of ws2801, each channel uses two pins, DATA on the first pin and CLOCK on the next. Only 24
channels of output are available and to reduce CPU usage, `opc-server` should be called with `--strip-count 24` or
lower.


Output Features
---------------
`opc-server` supports Fadecandy-inspired temporal dithering and interpolation
to enhance the smoothness of the output data. By default, it will apply a
luminance curve, interpolate and dither input data at the highest framerate
possible with the given number of LEDs.

These options can be configured by command-line switches that are documented in the help output from `opc-server -h`.

To disable all signal enhancements, use `opc-server -lut`

Demo Modes
--------------
`opc-server` supports several demo modes that will drive the attached pixels autonomously. This can help greatly with testing.

The demo mode is set using the `demoMode` parameter and can have the following values...

| demoMode | behavior
|----------|----------|
|none      |No demo running. Pixels will only update in response to incoming OPC packets|
|id        | Set the pixel to the strip index unless the pixel has the same index as the strip, then light it up grey with bit value: 1010 1010|
|fade      | Display a pleasing pattern of rotating color hues with a border that steps across the pixels every 12 seconds |
|black     | All pixels off|
|power     | All pixels on full white (based on current settings)- good for testing for maximum power requirements for current settings |

The default `demo-mode` set in the supplied `ws281x-config.json` configuration file is `fade`.

Note that received OPC data will override any currently running `demo-mode`.  The currently running `demo-mode` will resume display 5 seconds after the last OPC data is displayed. 

Configuration
====

Config info is typically stored in `/etc/ledscape-config.json`.

##Default config

The default config after installation is set up to drive WS281X strips connected to all of the 48 available output pins. Note that not all pins will work on BeagleBone Black unless you [disable the HDMI port](#HDMI Conflict).  

A description of the file format and some example configurations are available in the [`configs/` subdirectory](/configs) of this repo. 

##Directly editing the current config

You can edit the config file directly by typing...

```
nano /etc/ledscape-config.json
```

If the server is already running as a service, you'll need to enter the following command to get it to read the new config file...

```
sudo systemctl restart ledscape.service
```


Processing Examples
========

LEDscape provides versions of the FadeCandy processing examples modified to work better with LEDscape in the
`processing` directory. Clone this repo on a computer and run these sketches, edited to point at your BBB hostname or
ip address after starting `opc-server` or installing the system service.


Hardware Tips
========

Remember that the BBB outputs data at 3.3v. Depending on the specific LED strips, it is often possible to connect the `DI` on directly to one of the output pins on the BeagleBone, especially if the strips are high quality and the connecting wire is short. Many [recommend](https://forum.pjrc.com/threads/24648-Newbie-findings-re-WS281X-signal-quality(wire-length-resistors-and-grounds-Oh-my!) also adding an impedance  matching resistor to smooth out the signal.

If long cables (over ~6 feet) are used between the first pixel and the BeagleBone Black, issues with corruption of pixel data may arise as a result of the capacitance of the long cabling, as well as due to crosstalk between signal lines run in close proximity. While slighty increasing the impedance matching resistor value to 220 ohms will lessen these detrimental effects by limiting slew rate, the use of differential signaling is required for reliable operation with long cable runs. 

If your strips require 5V on `DI`, you will need to use a level-shifter of some sort. [Adafruit](http://www.adafruit.com/products/757) has a decent one which works well.  For custom circuit boards we recommend the [TI SN74LV245](http://octopart.com/partsearch#!?q=SN74LV245).

While there may be others, RGB123 makes an excellent 24/48 pin cape designed 
specifically for this version of LEDscape: [24 pin](http://rgb-123.com/product/beaglebone-black-24-output-cape/) or [48 pin](http://rgb-123.com/product/beaglebone-black-48-output-cape/) Be aware that these capes are NOT compatible with the x-r30 output modes!


API
===

`ledscape.h` defines the API. The key components are:

	ledscape_t * ledscape_init(unsigned num_pixels)
	ledscape_frame_t * ledscape_frame(ledscape_t*, unsigned frame_num);
	ledscape_draw(ledscape_t*, unsigned frame_num);
	unsigned ledscape_wait(ledscape_t*)

You can double buffer like this:

	const int num_pixels = 256;
	ledscape_t * const leds = ledscape_init(num_pixels);

	unsigned i = 0;
	while (1)
	{
		// Alternate frame buffers on each draw command
		const unsigned frame_num = i++ % 2;
		ledscape_frame_t * const frame
			= ledscape_frame(leds, frame_num);

		render(frame);

		// wait for the previous frame to finish;
		ledscape_wait(leds);
		ledscape_draw(leds, frame_num);
	}

	ledscape_close(leds);

The 24-bit RGB data to be displayed is laid out with BRGA format,
since that is how it will be translated during the clock out from the PRU.
The frame buffer is stored as a "strip-major" array of pixels.

	typedef struct {
		uint8_t b;
		uint8_t r;
		uint8_t g;
		uint8_t a;
	} __attribute__((__packed__)) ledscape_pixel_t;

	typedef struct {
		ledscape_pixel_t strip[32];
	} __attribute__((__packed__)) ledscape_frame_t;


Low level API
=============

If you want to poke at the PRU directly, there is a command structure
shared in PRU DRAM that holds a pointer to the current frame buffer,
the length in pixels, a command byte and a response byte.
Once the PRU has cleared the command byte you are free to re-write the
dma address or number of pixels.

	typedef struct
	{
		// in the DDR shared with the PRU
		const uintptr_t pixels_dma;

		// Length in pixels of the longest LED strip.
		unsigned num_pixels;

		// write 1 to start, 0xFF to abort. will be cleared when started
		volatile unsigned command;

		// will have a non-zero response written when done
		volatile unsigned response;
	} __attribute__((__packed__)) ws281x_command_t;

Reference
==========
* http://www.adafruit.com/products/1138
* http://www.adafruit.com/datasheets/WS2811.pdf
* http://processors.wiki.ti.com/index.php/PRU_Assembly_Instructions


debian/changelog
ledscape-ng (0.2.3-1) UNRELEASED; urgency=medium

  [ pi@arm-builder.home ]
  * Correct uio driver race on startup
  * Correct universe allocation to round properly

 --  <pi@arm-builder.home>  Wed, 14 Jun 2017 15:33:37 -0500

ledscape-ng (0.2-1) unstable; urgency=low

  * Initial release

 --  unknown <pi@arm-builder.home>  Tue, 30 May 2017 16:29:35 -0500


ledscape.h
/** \file
 * LEDscape for the BeagleBone Black.
 *
 * Drives up to 32 ws281x LED strips using the PRU to have no CPU overhead.
 * Allows easy double buffering of frames.
 */

#ifndef _ledscape_h_
#define _ledscape_h_

#include <stdint.h>
#include "pru.h"

/** The number of strips supported.
 *
 * Changing this also requires changes in ws281x.p to stride the
 * correct number of bytes per row..
 */
#define LEDSCAPE_NUM_STRIPS 48


/**
 * An LEDscape "pixel" consists of three channels of output and an unused fourth channel. The color mapping of these
 * channels is not defined by the pixel construct, but is specified by color_channel_order_t. Use ledscape_pixel_set_color
 * to assign color values to a pixel.
 */
typedef struct {
	uint8_t a;// was blue
	uint8_t b;// was red
	uint8_t c;// was green
	uint8_t d;// now used for white
} __attribute__((__packed__)) ledscape_pixel_t;


/** LEDscape frame buffer is "strip-major".
 *
 * All 32 strips worth of data for each pixel are stored adjacent.
 * This makes it easier to clock out while reading from the DDR
 * in a burst mode.
 */
typedef struct {
	ledscape_pixel_t strip[LEDSCAPE_NUM_STRIPS];
} __attribute__((__packed__)) ledscape_frame_t;

typedef struct ws281x_command ws281x_command_t;

typedef struct {
	ws281x_command_t * ws281x_0;
	ws281x_command_t * ws281x_1;
	pru_t * pru0;
	pru_t * pru1;
	const char* pru0_program_filename;
	const char* pru1_program_filename;
	unsigned num_pixels;
	size_t frame_size;
} ledscape_t;


typedef enum {
	COLOR_ORDER_RGB,
	COLOR_ORDER_RBG,
	COLOR_ORDER_GRB,
	COLOR_ORDER_GBR,
	COLOR_ORDER_BGR,
	COLOR_ORDER_BRG, // Old LEDscape default
	COLOR_ORDER_WBRG
} color_channel_order_t;


extern ledscape_t * ledscape_init(
unsigned num_pixels
);

extern ledscape_t * ledscape_init_with_programs(
	unsigned num_pixels,
	const char* pru0_program_filename,
	const char* pru1_program_filename
);


extern ledscape_frame_t *
ledscape_frame(
	ledscape_t * const leds,
	unsigned frame
);

extern void
ledscape_draw(
	ledscape_t * const leds,
	unsigned frame
);

inline void ledscape_pixel_set_color_rgbw(
	ledscape_pixel_t * const out_pixel,
	color_channel_order_t color_channel_order,
	uint8_t r,
	uint8_t g,
	uint8_t b,
	uint8_t w
) {
	switch (color_channel_order) {
		case COLOR_ORDER_RGB:
			out_pixel->a = r;
			out_pixel->b = g;
			out_pixel->c = b;
		break;
		case COLOR_ORDER_WBRG:
			out_pixel->a = w;
			out_pixel->b = b;
			out_pixel->c = r;
			out_pixel->d = g;
		break;
		case COLOR_ORDER_RBG:
			out_pixel->a = r;
			out_pixel->b = b;
			out_pixel->c = g;
		break;

		case COLOR_ORDER_GRB:
			out_pixel->a = g;
			out_pixel->b = r;
			out_pixel->c = b;
		break;

		case COLOR_ORDER_GBR:
			out_pixel->a = g;
			out_pixel->b = b;
			out_pixel->c = r;
		break;

		case COLOR_ORDER_BGR:
			out_pixel->a = b;
			out_pixel->b = g;
			out_pixel->c = r;
		break;

		case COLOR_ORDER_BRG:
			out_pixel->a = b;
			out_pixel->b = r;
			out_pixel->c = g;
		break;
	}
}

inline void ledscape_pixel_set_color(
	ledscape_pixel_t * const out_pixel,
	color_channel_order_t color_channel_order,
	uint8_t r,
	uint8_t g,
	uint8_t b
) {
	ledscape_pixel_set_color_rgbw(out_pixel, color_channel_order, r, g, b, 0);
}

inline void ledscape_set_color_rgbw(
	ledscape_frame_t * const frame,
	color_channel_order_t color_channel_order,
	uint8_t strip,
	uint16_t pixel,
	uint8_t r,
	uint8_t g,
	uint8_t b,
	uint8_t w
) {
	ledscape_pixel_set_color_rgbw(
		&frame[pixel].strip[strip],
		color_channel_order,
		r,
		g,
		b,
		w
	);
}

inline void ledscape_set_color(
	ledscape_frame_t * const frame,
	color_channel_order_t color_channel_order,
	uint8_t strip,
	uint16_t pixel,
	uint8_t r,
	uint8_t g,
	uint8_t b
) {
	ledscape_pixel_set_color(
		&frame[pixel].strip[strip],
		color_channel_order,
		r,
		g,
		b
	);
}

extern void
ledscape_wait(
	ledscape_t * const leds
);


extern void
ledscape_close(
	ledscape_t * const leds
);


extern const char* color_channel_order_to_string(
	color_channel_order_t color_channel_order
);

extern color_channel_order_t color_channel_order_from_string(
	const char* str
);

#endif


debian/watch.ex
# Example watch control file for uscan
# Rename this file to "watch" and then you can run the "uscan" command
# to check for upstream updates and more.
# See uscan(1) for format

# Compulsory line, this is a version 3 file
version=3

# Uncomment to examine a Webpage
# <Webpage URL> <string match>
#http://www.example.com/downloads.php ledscape-ng-(.*)\.tar\.gz

# Uncomment to examine a Webserver directory
#http://www.example.com/pub/ledscape-ng-(.*)\.tar\.gz

# Uncommment to examine a FTP server
#ftp://ftp.example.com/pub/ledscape-ng-(.*)\.tar\.gz debian uupdate

# Uncomment to find new files on sourceforge, for devscripts >= 2.9
# http://sf.net/ledscape-ng/ledscape-ng-(.*)\.tar\.gz

# Uncomment to find new files on GooglePages
# http://example.googlepages.com/foo.html ledscape-ng-(.*)\.tar\.gz


pru/build_template.sh
#!/usr/bin/env bash

cd $(dirname $0)

MAPPING_DIR=mappings
TEMPLATE_DIR=templates
OUTPUT_DIR=generated

PROGRAM_NAME=$1

echo "- Processing template $PROGRAM_NAME..."

for MAPPING_NAME in $(cd "$MAPPING_DIR"; ls *.json|sed s/.json$//); do
	echo "  - Building permutation for mapping $MAPPING_NAME"

	echo "    - Generating mapping headers..."
	nodejs pinmap.js $PROGRAM_NAME --mapping $MAPPING_NAME pru-headers > $OUTPUT_DIR/mapping-$MAPPING_NAME-p.h || exit -1

	echo "    - Applying template..."

	PRU0_FILENAME=$OUTPUT_DIR/$PROGRAM_NAME-$MAPPING_NAME-pru0.p
	echo "#define PRU_NUM 0" > $PRU0_FILENAME
	echo "#define PRU_NUM 0" > $PRU0_FILENAME
	echo "#include \"mapping-$MAPPING_NAME-p.h\"" >> $PRU0_FILENAME
	echo "#include \"../$TEMPLATE_DIR/$PROGRAM_NAME.p\"" >> $PRU0_FILENAME

	PRU1_FILENAME=$OUTPUT_DIR/$PROGRAM_NAME-$MAPPING_NAME-pru1.p
	echo "#define PRU_NUM 1" > $PRU1_FILENAME
	echo "#include \"mapping-$MAPPING_NAME-p.h\"" >> $PRU1_FILENAME
	echo "#include \"../$TEMPLATE_DIR/$PROGRAM_NAME.p\"" >> $PRU1_FILENAME

	echo
done


ledscape.c
/** \file
 * Userspace interface to the WS281x LED strip driver.
 *
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include <inttypes.h>
#include <errno.h>
#include <unistd.h>
#include "ledscape.h"


/** GPIO pins used by the LEDscape.
 *
 * The device tree should handle this configuration for us, but it
 * seems horribly broken and won't configure these pins as outputs.
 * So instead we have to repeat them here as well.
 *
 * If these are changed, be sure to check the mappings in
 * ws281x.p!
 *
 * See https://github.com/ehayon/BeagleBone-GPIO/blob/master/src/am335x.h
 * for a complete list of pins.
 *
 * TODO: Find a way to unify this with the defines in the .p file
 */
static const uint8_t gpios0[] = {
	2, 3, 7, 8, 9, 10, 11, 14, 20, 22, 23, 26, 27, 30, 31
};

static const uint8_t gpios1[] = {
	12, 13, 14, 15, 16, 17, 18, 19, 28, 29
};

static const uint8_t gpios2[] = {
	1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 22, 23, 24, 25,
};

static const uint8_t gpios3[] = {
	14, 15, 16, 17, 19, 21
};

#define ARRAY_COUNT(a) ((sizeof(a) / sizeof(*a)))


/** Command structure shared with the PRU.
 *
 * This is mapped into the PRU data RAM and points to the
 * frame buffer in the shared DDR segment.
 *
 * Changing this requires changes in ws281x.p
 */
typedef struct ws281x_command
{
	// in the DDR shared with the PRU
	uintptr_t pixels_dma;

	// Length in pixels of the longest LED strip.
	unsigned num_pixels;

	// write 1 to start, 0xFF to abort. will be cleared when started
	volatile unsigned command;

	// will have a non-zero response written when done
	volatile unsigned response;
} __attribute__((__packed__)) ws281x_command_t;


/** Retrieve one of the two frame buffers. */
ledscape_frame_t *
ledscape_frame(
	ledscape_t * const leds,
	unsigned int frame
)
{
	if (frame >= 2)
		return NULL;

	return (ledscape_frame_t*)((uint8_t*) leds->pru0->ddr + leds->frame_size * frame);
}


/** Initiate the transfer of a frame to the LED strips */
void
ledscape_draw(
	ledscape_t * const leds,
	unsigned int frame
)
{

	leds->ws281x_0->pixels_dma = leds->pru0->ddr_addr + leds->frame_size * frame;
	leds->ws281x_1->pixels_dma = leds->pru0->ddr_addr + leds->frame_size * frame;

	// Wait for any current command to have been acknowledged
	while (leds->ws281x_0->command || leds->ws281x_1->command);

	// Zero the responses so we can wait for them
	leds->ws281x_0->response = leds->ws281x_1->response = 0;

	// Send the start command
	leds->ws281x_0->command = 1;
	leds->ws281x_1->command = 1;
}


/** Wait for the current frame to finish transfering to the strips.
 * \returns a token indicating the response code.
 */
void
ledscape_wait(
	ledscape_t * const leds
)
{
	while (1)
	{
		pru_wait_interrupt();

		// printf("pru0: (%d,%d), pru1: (%d,%d)\n",
		// 	leds->ws281x_0->command, leds->ws281x_0->response,
		// 	leds->ws281x_1->command, leds->ws281x_1->response
		// );

		if (leds->ws281x_0->response && leds->ws281x_1->response) return;
	}
}


ledscape_t * ledscape_init( unsigned num_pixels ) {
	return ledscape_init_with_programs(
		num_pixels,
		"pru/bin/ws281x-original-ledscape-pru0.bin",
		"pru/bin/ws281x-original-ledscape-pru1.bin"
	);
}

ledscape_t * ledscape_init_with_programs(
	unsigned num_pixels,
	const char* pru0_program_filename,
	const char* pru1_program_filename
)
{
	pru_t * const pru0 = pru_init(0);
	pru_t * const pru1 = pru_init(1);

	const size_t frame_size = num_pixels * LEDSCAPE_NUM_STRIPS * 4;

	if (2*frame_size > pru0->ddr_size)
		die("Pixel data needs at least 2 * %zu, only %zu in DDR\n",
			frame_size,
			pru0->ddr_size
		);

	ledscape_t * const leds = calloc(1, sizeof(*leds));

	*leds = (ledscape_t) {
		.pru0		= pru0,
		.pru1		= pru1,
		.num_pixels	= num_pixels,
		.frame_size	= frame_size,
		.pru0_program_filename  = pru0_program_filename,
		.pru1_program_filename  = pru1_program_filename,
		.ws281x_0	= pru0->data_ram,
		.ws281x_1	= pru1->data_ram
	};

	*(leds->ws281x_0) = *(leds->ws281x_1) = (ws281x_command_t) {
		.pixels_dma	= 0, // will be set in draw routine
		.command	= 0,
		.response	= 0,
		.num_pixels	= leds->num_pixels,
	};

	// Configure all of our output pins.
	/* Device tree works now! Not needed.

	for (unsigned i = 0 ; i < ARRAY_COUNT(gpios0) ; i++)
		pru_gpio(0, gpios0[i], 1, 0);
	for (unsigned i = 0 ; i < ARRAY_COUNT(gpios1) ; i++)
		pru_gpio(1, gpios1[i], 1, 0);
	for (unsigned i = 0 ; i < ARRAY_COUNT(gpios2) ; i++)
		pru_gpio(2, gpios2[i], 1, 0);
	for (unsigned i = 0 ; i < ARRAY_COUNT(gpios3) ; i++)
		pru_gpio(3, gpios3[i], 1, 0);
	*/
	// Initiate the PRU0 program
	pru_exec(pru0, pru0_program_filename);

	// Watch for a done response that indicates a proper startup
	// \todo timeout if it fails
	fprintf(stdout, "String PRU0 with %s... ", pru0_program_filename);
	while (!leds->ws281x_0->response);
	printf("OK\n");


	// Initiate the PRU1 program
	pru_exec(pru1, pru1_program_filename);

	// Watch for a done response that indicates a proper startup
	// \todo timeout if it fails
	fprintf(stdout, "String PRU1 with %s... ", pru1_program_filename);
	while (!leds->ws281x_1->response);
	printf("OK\n");

	return leds;
}


const char* color_channel_order_to_string(color_channel_order_t color_channel_order) {
	switch (color_channel_order) {
		case COLOR_ORDER_RGB: return "RGB";
		case COLOR_ORDER_WBRG: return "WBRG";
		case COLOR_ORDER_RBG: return "RBG";
		case COLOR_ORDER_GRB: return "GRB";
		case COLOR_ORDER_GBR: return "GBR";
		case COLOR_ORDER_BGR: return "BGR";
		case COLOR_ORDER_BRG: return "BRG";
		default: return  "<invalid color_channel_order>";
	}
}

color_channel_order_t color_channel_order_from_string(const char* str) {
	if (strcasecmp(str, "RGB") == 0) {
		return COLOR_ORDER_RGB;
	}
	else if (strcasecmp(str, "RBG") == 0) {
		return COLOR_ORDER_RBG;
	}
	else if (strcasecmp(str, "GRB") == 0) {
		return COLOR_ORDER_GRB;
	}
	else if (strcasecmp(str, "GBR") == 0) {
		return COLOR_ORDER_GBR;
	}
	else if (strcasecmp(str, "BGR") == 0) {
		return COLOR_ORDER_BGR;
	}
	else if (strcasecmp(str, "BRG") == 0) {
		return COLOR_ORDER_BRG;
	}
	else if (strcasecmp(str, "WBRG") == 0) {
		return COLOR_ORDER_WBRG;
	}
	else {
		return -1;
	}
}

void
ledscape_close(
	ledscape_t * const leds
)
{
	// Signal a halt command
	leds->ws281x_0->command = 0xFF;
	leds->ws281x_1->command = 0xFF;
	pru_close(leds->pru0);
	pru_close(leds->pru1);
}
